homeassistant/components/screenlogic/sensor.py

```diff
@@ -85,190 +158,121 @@
     STATE_TYPE.TOTAL_INCREASING: SensorStateClass.TOTAL_INCREASING,
 }
 
-SL_UNIT_TO_HA_UNIT = {
-    UNIT.CELSIUS: UnitOfTemperature.CELSIUS,
-    UNIT.FAHRENHEIT: UnitOfTemperature.FAHRENHEIT,
-    UNIT.MILLIVOLT: UnitOfElectricPotential.MILLIVOLT,
-    UNIT.WATT: UnitOfPower.WATT,
-    UNIT.HOUR: UnitOfTime.HOURS,
-    UNIT.SECOND: UnitOfTime.SECONDS,
-    UNIT.REVOLUTIONS_PER_MINUTE: REVOLUTIONS_PER_MINUTE,
-    UNIT.PARTS_PER_MILLION: CONCENTRATION_PARTS_PER_MILLION,
-    UNIT.PERCENT: PERCENTAGE,
-}
-
 
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up entry."""
-    entities: list[ScreenLogicSensorEntity] = []
-    coordinator: ScreenlogicDataUpdateCoordinator = hass.data[DOMAIN][
+    entities: list[ScreenLogicSensor] = []
+    coordinator: ScreenlogicDataUpdateCoordinator = hass.data[SL_DOMAIN][
         config_entry.entry_id
     ]
-    equipment_flags = coordinator.gateway_data[SL_DATA.KEY_CONFIG]["equipment_flags"]
-
-    # Generic push sensors
-    for sensor_name in coordinator.gateway_data[SL_DATA.KEY_SENSORS]:
-        if sensor_name in SUPPORTED_BASIC_SENSORS:
-            entities.append(
-                ScreenLogicStatusSensor(coordinator, sensor_name, CODE.STATUS_CHANGED)
-            )
+    gateway = coordinator.gateway
+    data_path: ScreenLogicDataPath
+    value_params: SupportedSensorValueParameters
+    for data_path, value_params in iterate_expand_group_wildcard(
+        gateway, SUPPORTED_DATA
+    ):
+        entity_key = generate_unique_id(*data_path)
+
+        device = data_path[0]
+
+        if not (DEVICE_INCLUSION_RULES.get(device) or value_params.included).test(
+            gateway, data_path
+        ):
+            cleanup_excluded_entity(coordinator, DOMAIN, entity_key)
+            continue
+
+        try:
+            value_data = gateway.get_data(*data_path, strict=True)
+        except KeyError:
+            _LOGGER.debug("Failed to find %s", data_path)
+            continue
+
+        entity_kwargs = {
+            "data_path": data_path,  #
+            "key": entity_key,  #
+            "entity_category": value_params.entity_category,
+            "entity_registry_enabled_default": value_params.enabled.test(
+                gateway, data_path
+            ),
```

Comment:
You could do something like `**build_base_entity_description(gateway, value_params, data_path)` to still keep it a bit more DRY
for the items in https://github.com/home-assistant/core/blob/6cd28b64e8b9f6af833ea5eb267c4d29fed6cccb/homeassistant/helpers/entity.py#L197

Reply:
I ended up needing to pass `entity_key` and `value_data` as well.

I think I could get to `**build_base_entity_description(gateway, value_params)` by packing those into `value_params`, including `data_path`, and _only_ iterating on `value_params`, but I was worried that would start to get harder to follow. It would require some changes to `iterate_expand_group_wildcard()` as well as I would probably generate the `entity_key` there.